python/mpt/src/mpt/trie_diff.cairo

```diff
@@ -278,34 +515,1092 @@ func _process_storage_diff{
+
+    // path = path + bytes([i])
+    let path_copy = Bytes__copy__(path);
+    assert path_copy.value.data[path_copy.value.len] = index;
+    tempvar sub_path = Bytes(
+        new BytesStruct(data=path_copy.value.data, len=path_copy.value.len + 1)
+    );
+
+    let first_nib = right.value.rest_of_key.value.data[0];
+    if (first_nib == index) {
+        // Compare to the shortened leaf node
+        tempvar right_leaf_shortened = LeafNode(
+            new LeafNodeStruct(
+                rest_of_key=Bytes(
+                    new BytesStruct(
+                        data=right.value.rest_of_key.value.data + 1,
+                        len=right.value.rest_of_key.value.len - 1,
+                    ),
+                ),
+                value=right.value.value,
+            ),
+        );
+        let right_shortened = OptionalUnionInternalNodeExtendedImpl.from_leaf(right_leaf_shortened);
+        _compute_diff(
+            left=subnode_i, right=right_shortened, path=sub_path, account_address=account_address
+        );
+        return _compute_left_branch_on_right_leaf(
+            left=left, right=right, path=path, account_address=account_address, index=index + 1
+        );
+    }
+    // Compare to None
+    tempvar null_node = OptionalUnionInternalNodeExtended(
+        cast(0, OptionalUnionInternalNodeExtendedEnum*)
+    );
+    _compute_diff(left=subnode_i, right=null_node, path=sub_path, account_address=account_address);
+
+    return _compute_left_branch_on_right_leaf(
+        left=left, right=right, path=path, account_address=account_address, index=index + 1
+    );
+}
+
+func _compute_left_branch_on_right_extension_node{
+    range_check_ptr,
+    bitwise_ptr: BitwiseBuiltin*,
+    poseidon_ptr: PoseidonBuiltin*,
+    node_store: NodeStore,
+    address_preimages: MappingBytes32Address,
+    storage_key_preimages: MappingBytes32Bytes32,
+    main_trie_end: AddressAccountNodeDictAccess*,
+    storage_tries_end: TupleAddressBytes32U256DictAccess*,
+}(
+    left: BranchNode,
+    right: ExtensionNode,
+    path: Bytes,
+    account_address: OptionalAddress,
+    index: felt,
+) -> () {
+    alloc_locals;
+
+    if (index == 16) {
+        return ();
+    }
+
+    let subnodes_ptr = cast(left.value.subnodes.value, felt*);
+    let subnode_i_extended = Extended(cast(subnodes_ptr[index], ExtendedEnum*));
+    let subnode_i = OptionalUnionInternalNodeExtendedImpl.from_extended(subnode_i_extended);
+
+    // path = path + bytes([i])
+    let path_copy = Bytes__copy__(path);
+    assert path_copy.value.data[path_copy.value.len] = index;
+    tempvar sub_path = Bytes(
+        new BytesStruct(data=path_copy.value.data, len=path_copy.value.len + 1)
+    );
+
+    let first_nib = right.value.key_segment.value.data[0];
+    if (first_nib == index) {
+        // Compare to the shortened extension node
+        tempvar extension = ExtensionNode(
+            new ExtensionNodeStruct(
+                key_segment=Bytes(
```

Comment:
here we're sure that an extension node should have key_segment length gte 1 so it's ok